### PR TITLE
CThread: create IThreadImpl to improve platform abstraction visibility

### PR DESCRIPTION
**Pull request by lrusak**
This PR continues the cleanup of `CThread`. Before this PR we used some ugly includes to pull in code that required ifdeffing.

This PR intends to abstract the platform threading implementation out of `CThread`. This is done with a new class `IThreadImpl`. This interface is then used by each platform specific class. After all the cleanup in #20883 the platform implementation only needs to be able to set the thread priority. Setting the priority isn't possible using `std::thread` and is thus why this platform implementation existed in the first place.

The idea here is to call a static method (there can only be one) to create the platform specific implementation class. This class is then used when calling `CThread::SetPriority`.

After this the platform specific code can be moved to the `platform/*/threads` directory. This also allows platforms to more easily create specific implementations if necessary. I've already done these changes but I'm PRing in smaller chunks to keep the review burden down.

I've also included a couple other commits:
 - unistd.h include seems to be needed as the build fails with some of the header shuffling.
 - run clang-format on RecursiveMutex.h as it has non standard indentation.
 - Move the CRecursiveMutex code from `threads/platform/pthreads/ThreadImpl.cpp` to it's own cpp file.

### CODE REVIEW

File: xbmc/threads/platform/RecursiveMutex.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#endif
+    alreadyCalled = true;
+  }
+
+  return true; // note, we never call destroy.
+}
+
+static bool recursiveAttrSet = SetRecursiveAttr();
+
+pthread_mutexattr_t* CRecursiveMutex::getRecursiveAttr()
```
Comment (garbear) [inline xbmc/threads/platform/RecursiveMutex.cpp:35]: I know this signature comes from existing code, but is there any merit to returning `pthread_mutexattr_t&` to indicate that `nullptr` can't be returned?

**Comment by garbear** [comment]
I checked the details, and found two small things. I don't have a handle on the larger picture, but "create IThreadImpl to improve platform abstraction visibility" is an admirable goal and I assume it was done correctly.

File: xbmc/threads/Thread.h [hunk 2]
```diff
@@ -73,15 +71,13 @@ class CThread
     return std::this_thread::get_id();
   }
 
-  // -----------------------------------------------------------------------------------
-  // These are platform specific and can be found in ./platform/[platform]/ThreadImpl.cpp
-  // -----------------------------------------------------------------------------------
-
-  // Get and set the thread's priority
+  /**
```
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/threads/Thread.h:74]: There are a few different styles I guess, but it our guidelines require `/*!`.

I was using the vscode doxygen generator https://marketplace.visualstudio.com/items?itemName=cschlosser.doxdocgen

I guess I'll have to adjust the settings to make it conform to my style.

Thanks for pointing this out.

File: xbmc/threads/platform/RecursiveMutex.cpp (hunk 1 above)
Comment (lrusak) [reply xbmc/threads/platform/RecursiveMutex.cpp]: I added this change as another commit so it's obvious what was changed.

**Review by garbear — APPROVED** [review]

**Pull request merged**